AP_Motors: Add a seprate spool down time, if it's 0 use spool up time

WickedShell opened the pull request:
This relates to #20354 and allows you to have a separate spool down timer which can be much shorter then spooling up. If it's set to 0 then we will use the spool up time.

As usual with parameter names, I'm open to suggestions, there just wasn't quite enough space to get the name in there well.

This was tested with a scope on the motor outputs, and parameter manipulation.

IamPete1 reviewed (commented):
This still allows the throttle to go up in the spool down state until `_throttle_thrust_max` comes down enough to limit it.

If we were to fix that would we still need this separate rate?

I was thinking we could just set `_throttle_thrust_max` to the motors throttle output level when we enter `SpoolState::SPOOLING_DOWN`.

This is originally a plane issue, my understanding is that this change won't actually do anything on plane with https://github.com/ArduPilot/ardupilot/pull/20355 ?

lthall commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -553,9 +563,9 @@ void AP_MotorsMulticopter::output_logic()
     }
 
-    if (_spool_up_time < 0.05) {
+    if (_spool_up_time < minimum_spool_time) {
```
Our normal convention is to use = 0 to turn off a parameter. We may have to remove the _spool_up_time.set(minimum_spool_time) and just use the minimum value if it is zero.
That way if we use _spool_down_time = 0 we can use _spool_up_time.

WickedShell replied:
This part isn't a change of behavior at the moment. I'm not opposed to changing this, but I think it's somewhat fair that this is actually changing the parameter and the user will have a chance to see see it when they download parameters. 

Spool up time was previous and had to be enabled. Are you saying you want a change in behavior on spool up time?

lthall replied:
Yes, I agree.

lthall reviewed (commented):
I think the mechanics are all correct.

The only question I have is should we update the = 0 parameter handling...

WickedShell commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -537,6 +546,7 @@ void AP_MotorsMulticopter::update_throttle_hover(float dt)
 // run spool logic
 void AP_MotorsMulticopter::output_logic()
 {
+    const constexpr float minimum_spool_time = 0.05f;
```
Sure, not at all bothered, I was just preserving the old value, but I'm happy to change it.

lthall replied:
I am happy to keep this value.

WickedShell commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -207,6 +207,15 @@ const AP_Param::GroupInfo AP_MotorsMulticopter::var_info[] = {
 
+    // @Param: SPOOL_TIM_DN
+    // @DisplayName: Spool down time
+    // @Description: Time taken to spool down the motors from min to zero throttle. If set to 0 then SPOOL_TIME is used instead.
```
Copy paste error on my part, good catch.

WickedShell replied:
Actually I take that back. It's primarily meant for min -> zero, not max -> zero. It's really just the spool down from throttle unlimited to zero, but it's normally at a low throttle when that starts. I think adding the phrase max to zero implies a general slew limit, which this isn't the right parameter for.

IamPete1 reviewed (approved):
LGTM

tridge commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -597,34 +607,39 @@ void AP_MotorsMulticopter::output_logic()
-        case DesiredSpoolState::GROUND_IDLE:
+        }
+        case DesiredSpoolState::GROUND_IDLE: {
+            const float spool_step = 1.0f / (_spool_up_time * _loop_rate);
```
shouldn't this check the current state to see if its going up or down?

WickedShell replied:
Ah, okay, I wasn't tracking that state, I was mostly thinking about the complete on/off. I can add a check for that.

WickedShell replied:
Done. Please double check that though.

WickedShell commented:
Rebased, addressed tridge's concern. I've flown this briefly in SITL, and the stuck it on a real copter with an absurd 18 second spool up, and 4 second spool down and got the expected behaviour on the ground with no props on the aircraft.

lthall commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -656,7 +677,9 @@ void AP_MotorsMulticopter::output_logic()
         _spin_up_ratio = 1.0f;
-        _throttle_thrust_max -= spool_step;
+        const float spool_time = _spool_down_time > minimum_spool_time ? _spool_down_time : _spool_up_time;
+        const float spool_adjustment = -1.0f / (spool_time * _dt);
```
why is this equation different. Shouldn't it be:
spool_step = _dt / _spool_up_time;

WickedShell replied:
Victim of a bad rebase, fixed it.

lthall commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -656,7 +677,9 @@ void AP_MotorsMulticopter::output_logic()
-        _throttle_thrust_max -= spool_step;
+        const float spool_time = _spool_down_time > minimum_spool_time ? _spool_down_time : _spool_up_time;
+        const float spool_adjustment = -1.0f / (spool_time * _dt);
+        _throttle_thrust_max += spool_adjustment;
```
I like:
_throttle_thrust_max -= spool_step;

As it is clear things are being reduced. Is there a reason to change this?

WickedShell replied:
There isn't anymore. At an earlier form I had kept `spool_step` as precomputed, and was avoiding aliasing, but at this point we aren't doing that anymore, so I've swapped it back.

lthall commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -668,7 +691,7 @@ void AP_MotorsMulticopter::output_logic()
         }
 
-        _thrust_boost_ratio = MAX(0.0, _thrust_boost_ratio - spool_step);
+        _thrust_boost_ratio = MAX(0.0, _thrust_boost_ratio + spool_adjustment);
```
This line shouldn't need to change if week keep the variable name spool_step
